From 7cc02009f07d000b708eff70f4f6cbb2b1060fcd Mon Sep 17 00:00:00 2001
From: Seth Schoen <schoen@loyalty.org>
Date: Mon, 14 Jun 2021 21:29:06 -0700
Subject: [PATCH] Don't use lowest address as broadcast (sysctl gated)

There are still other things to check, but this is a first cut
at making the lowest address no longer be treated as a broadcast.

It achieved interoperability on a LAN with an analogously-patched
Linux host (both could ARP/ping/netcat with each other when either
was using the .0 address on a /24).

Introduces new sysctl: net.inet.ip.subnet_addr_is_broadcast
(boolean: 0 means subnet address itself, as host address is not
treated as a broadcast address; 1 means that it is, the
traditional behavior specified by current RFCs).
---
 sys/netinet/in.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/sys/netinet/in.c b/sys/netinet/in.c
index bcf071a81e0e..bbd11f419426 100644
--- a/sys/netinet/in.c
+++ b/sys/netinet/in.c
@@ -81,6 +81,12 @@ static void	in_purgemaddrs(struct ifnet *);
 
 static bool	ia_need_loopback_route(const struct in_ifaddr *);
 
+static bool	subnet_addr_is_broadcast = false;
+
+SYSCTL_BOOL(_net_inet_ip, OID_AUTO, subnet_addr_is_broadcast, CTLFLAG_RW,
+    &subnet_addr_is_broadcast, false,
+    "Treat IPv4 subnet address as a broadcast address");
+
 VNET_DEFINE_STATIC(int, nosameprefix);
 #define	V_nosameprefix			VNET(nosameprefix)
 SYSCTL_INT(_net_inet_ip, OID_AUTO, no_same_prefix, CTLFLAG_VNET | CTLFLAG_RW,
@@ -1135,11 +1141,14 @@ in_ifaddr_broadcast(struct in_addr in, struct in_ifaddr *ia)
 
 	return ((in.s_addr == ia->ia_broadaddr.sin_addr.s_addr ||
 	     /*
-	      * Check for old-style (host 0) broadcast, but
-	      * taking into account that RFC 3021 obsoletes it.
+	      * Check for old-style (host 0) broadcast only if not
+	      * preempted by RFC 3021. Also gated by
+	      * subnet_addr_is_broadcast sysctl.
 	      */
-	    (ia->ia_subnetmask != IN_RFC3021_MASK &&
-	    ntohl(in.s_addr) == ia->ia_subnet)) &&
+           (ia->ia_subnetmask != IN_RFC3021_MASK &&
+           ntohl(in.s_addr) == ia->ia_subnet &&
+           subnet_addr_is_broadcast)) &&
+
 	     /*
 	      * Check for an all one subnetmask. These
 	      * only exist when an interface gets a secondary
-- 
2.25.1

